---
title: "All-female beetles and Wolbachia"
lightbox: false
draft: true
format: 
    closeread-html:
        remove-header-space: true
execute: 
  echo: false
  warning: false
  message: false
  output: false
---

<!-- New version of closeread somehow does not accept css in yaml header anymore. Just walking around it. -->

```{=html}
<head>
<link rel="stylesheet" href="/theme/closeread/closeread.css">
</head>
```

```{r}
#| label: setup
library(tidyverse)
library(readxl)
library(ggpubr)
library(zhehaoColors)
library(magrittr)
```

:::{.epigraph}
# All-female beetles and Wolbachia

Sometimes, biological research can be difficult to understand, and also difficult to explain, especially for those who don't work in the related fields. Even for people with proficiency under the same roof "Biology", the topics can also be completely poles apart. 

Here, I will try to use easy terms, and method called "scrollytelling", to tell you what I did in my bachelor thesis. 

:::

:::{.cr-section}
<br>
Imaging, in a sunny day at spring, we went for a comfortable hike outside Hamburg ... [@cr-meadow]

:::{#cr-meadow}
![](https://www.elbetal-mv.de/fileadmin/_processed_/b/8/csm_Brodarer_Brack_Elbe_A_Hollerbach_2a424fdcbc.jpg){width="350px"}
:::

There was a river flowing through the meadow, the weather has been warmed up for few weeks. We saw a tall bush by the river, with some beautiful pink flowers blooming on it. @cr-epilobium

:::{#cr-epilobium}
![](https://i0.wp.com/jeremybartlett.co.uk/wp-content/uploads/2022/08/aP8010050.jpg?qw=1000&ssl=1){width="350px"}
:::

And when we look closer ...

@cr-altica-wild We saw some dark, shiny little beetles sitting on the leaves. 

:::{#cr-altica-wild}
![](https://inaturalist-open-data.s3.amazonaws.com/photos/41099607/original.jpeg){width="350px"}
:::

We looked into the literature, it turns out this beetle should be *Altica lythri*, belongs to a family of beetles who are absolute experts of jumping. @cr-nadein2016

:::{#cr-nadein2016}
![](ba-scrolly-assets/alticini-takingoff-nadein2016.png){width="350px"}
:::

We wanted to know more about them, so we took some beetles back to the lab and tested their genetic identity with DNA-Barcoding. Which means a specific gene section was compared within the group, and with related species. @cr-dnabarcoding

:::{#cr-dnabarcoding}
![](https://upload.wikimedia.org/wikipedia/commons/thumb/b/bd/DNA_Barcoding.png/800px-DNA_Barcoding.png){width="350px"}
:::

But, it tells us something unexpected ...

Normally, if we talk about a **monophyletic** group (what a "species" should be), @cr-mammals

we will expect their genes to be very similar within the group, coming from the same ancestor ... [@cr-mammals]{pan-to="-40%, 50%" scale-by="1.7" data-crop="31%, 0%, 0%, 70%"}

And show a distinct difference towards others. [@cr-mammals]{pan-to="0%, -40%" scale-by="1" data-crop="0%, 63%, 0%, 0%"}

:::{#cr-mammals}
![](/closeread/ba-scrolly-assets/mammals.png)
<!-- https://www.researchgate.net/publication/6594089/figure/fig4/AS:341632844091400@1458463025457/ML-Phylogenetic-Tree-of-Mammals-The-topology-and-branch-lengths-shown-here-are-based-on.png -->
:::

In our case, *A. lythri* (**L**) forms multiple trees with other related species. We call such subgroups of one species **haplotypes**. [@jackel2013]{style="color: grey; font-size=10px"} @cr-jackel2013

:::{#cr-jackel2013}
![](ba-scrolly-assets/jackel2013.png)
:::

[@cr-jackel2013]{data-crop="45%, 26%, 30%, 45%"} Such as the HT1 ... 

[@cr-jackel2013]{data-crop="45%, 0%, 30%, 79%"} A similar, but different haplotype, HT1* ... 

[@cr-jackel2013]{data-crop="0%, 25%, 65%, 23%"} HT2 ... 

[@cr-jackel2013]{data-crop="40%, 73%, 35%, 0%"} and HT3 ... 

@cr-jackel2013 Evolutionary speaking, these haplotypes have different histories, we have to treat them as different. So we categorised them by their haplotypes, and investigated each group further. 

Then we found out something unusual ...

@cr-sexdis In HT1 and HT1*, there are only females. 

:::{#cr-sexdis}
```{r}
#| fig-width: 1
#| fig-height: 1
n <- 20
ht1st <- data.frame(id = 1:n, sex = factor(replicate(n, sample(c('female'), 1))), haplotype = factor("HT1*"))
ht1 <- data.frame(id = 1:n, sex = factor(replicate(n, sample(c('female'), 1))), haplotype = factor("HT1"))
ht3 <- data.frame(id = 1:n, sex = factor(replicate(n, sample(c('female', 'male'), 1))), haplotype = factor("HT3"))
ht2 <- data.frame(id = 1:n, sex = factor(replicate(n, sample(c('female', 'male'), 1))), haplotype = factor("HT2"))

x <- bind_rows(ht1st, ht1, ht2, ht3)

sexdis <- x %>%
    ggplot(aes(x = haplotype, fill = sex)) +
    geom_bar(position = "fill", stat = "count") +
    scale_y_continuous(labels = scales::percent_format()) +
    labs(x = "Haplotype", y = "", fill = "Sex") +
    zhehaoplot_transparent() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
        aspect.ratio = 1.3) + 
    use_palette_ggplot("copenhagen1")
    
ggsave("closeread/ba-scrolly-assets/sexdis.png", sexdis, width = 8, height = 8, units = "cm")
```

![](ba-scrolly-assets/sexdis.png)
:::

@cr-sexdis-n Considering the sample size we have, the chance that we haven't encounter any male individuals, but the true sex ratio be 1:1 on the field, is extremely low. 

:::{#cr-sexdis-n}
```{r}
sexdis_n <- sexdis +
    stat_count(
        geom = "text",
        aes(label = after_stat(count)),
        position = position_fill(vjust = 0.5),
        color = "white",
        size = 3.5)

ggsave("closeread/ba-scrolly-assets/sexdis-n.png", sexdis_n, width = 8, height = 8, units = "cm")
```

![](ba-scrolly-assets/sexdis-n.png)
:::

In other words, there must be a significant cause for this unusual sex ratio.

@cr-wol We did some digging and found out that there's a bacterium, called ***Wolbachia***, being known for the ability to push the sex ratio of its insect host towards female. 

:::{#cr-wol}
![](ba-scrolly-assets/Wolbachia.png)
:::

Interesting! Maybe we should also take a look at the *Wolbachia* infection of these beetles ...

:::{#cr-poolseq-tf}
```{r}
thesis_data <- read_xlsx("confidential/bachelor-thesis-data.xlsx", sheet = "Pool Screen", range = "A4:O94")
thesis_data %<>% mutate(infected = `Infection type` != "Negative") %>%
    mutate(Haplotype = case_when(
        Haplotype == "HT1st" ~ "HT1*",
        Haplotype == "HT2/2st" ~ "HT2",
        .default = Haplotype
    )) %>%
    mutate(Haplotype = factor(Haplotype, levels = c("HT1*", "HT1", "HT2", "HT3"))) %>%
    filter(!is.na(Haplotype))
infect <- thesis_data %>%
    group_by(Haplotype) %>%
    ggplot() +
    geom_bar(aes(x = Haplotype, fill = infected), position = "fill", stat = "count") +
    scale_y_continuous(labels = scales::percent_format()) +
    labs(x = "Haplotype", y = "", fill = "Infection") +
    zhehaoplot_transparent() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
        aspect.ratio = 1.3) + 
    use_palette_ggplot("aaaaa") +
    theme(legend.position = "right", legend.direction = "vertical")
ggsave("closeread/ba-scrolly-assets/infectTF.png", infect, width = 8, height = 8, units = "cm")
```

![](ba-scrolly-assets/infectTF.png)
:::

[@cr-poolseq-tf]{data-crop="54%, 10%, 0%, 0%"} It turns out some HT3 are infected ... 

[@cr-poolseq-tf]{data-crop="43%, 10%, 0%, 0%"} HT2 are heavily infected ... 

[@cr-poolseq-tf]{data-crop="32%, 10%, 0%, 0%"} HT1 too ... 

[@cr-poolseq-tf]{data-crop="0%, 10%, 0%, 0%"} but HT1* are not quite a fan of *Wolbachia* 

Wait, this doesn't explain things very well.

[@cr-poolseq-tf]{data-crop="0%, 10%, 58%, 0%"} HT1 and HT1* are all-female, but they are not all infected. 

[@cr-poolseq-tf]{data-crop="43%, 10%, 0%, 0%"} HT2 and HT3 have normal males, but HT2 are heavily infected.

It turns out *Wolbachia* also has different types. We call them **strains**.



Most HT3 are infected by strains wA2 or strain wB.

Most HT2 are infected by both wA2 and wB.

All HT1 are infected by wA1.

Which is also the main strain living with HT1*.

:::
